### ISSUE TITLE
Feature: Molecule Serialization

### ISSUE THREAD

**Issue by shinzlet** [atomCAD/atomCAD#23]
In the long term, our serialization needs will likely be quite complex. To scratch the surface:
- Storing molecules & feature lists
- Storing assemblies (including subassemblies, which may mean referencing / embedding other files)
- Storing atom coordinate checkpoints every N features (still unclear if we need / want this)
- Storing time-series molecular animation / simulation results
- Saving metadata (author, creation time, etc)
- Conceivably storing a bunch of other things (copyright info, user-defined features via scripting, ...)

It's hard to implement or even think about most of these, as atomCAD is still very young few of those features are even being drafted yet.

To give us a starting-off point, I'm going to start researching the best ways of doing serialization in rust (I'm guessing serde will be the answer). Then, I plan to get serialization working for individual molecules. In the future, we won't even have a filetype for a single atomCAD molecule (`Assembly` is the type that represents a workspace), but in atomCAD's current state this is all we need.

Goals:
- It should be easy to create a FeatureList with a script (this allows easy experimentation with the modelling language being discussed)
- Files should be small (i.e atom coords should be stored as binary blobs, eventually this probably means using compression / diffs for storing time series data)
- The file format should be well documented
- The file format and serialization process should be flexible enough to support drastic future expansion
- The atomCAD file extension should look and sound cool (ok, that one is kind of *my* goal..)

I'll be working over in https://github.com/shinzlet/atomCAD/tree/serialization.

**Comment by maaku** [comment]
This may be a nit, but better to correct now while it is easiest. “Feature” for a saved undo tree is a bit confusing. Maybe an “Operation” which is an instance of Feature applied to a Model? That way we’re storing a history of operations applied to our model over time, while each operation is some feature of the software. That’s clearer nomenclature, I think.

**Comment by shinzlet** [comment]
Fair point! I just used fusion 360's naming convention without thinking about it, but you're right that it's a bit unclear.

**Comment by shinzlet** [comment]
After taking some time to experiment around, I've found a serialization issue that's actually pretty hard to get a good compromise on. 

Currently, operations are represented by structs that implement the trait `Feature`. This is useful because it allows us to encapsulate each operation quite well: its data (like what geometry it should create, what target atoms it anchors to, etc) is kept in some struct, and that struct implements `Feature` to expose an `apply` method. This is also easily extensible - if we later want to allow feature parameters to be edited in the UI, we can add more methods to `Feature`, like `create_control_panel`.

The important implication of this is that the `FeatureList` doesn't know what it will contain at compile time. That is why a feature list stores a `Vec<dyn Feature>`.

Serializing a `Vec<dyn Feature>` is not easy in Serde, but it is conceptually possible. What's really difficult is deserializing: Serde doesn't know what structs implement Feature, so when it finds an "AtomFeature" in the file it's reading, it won't have a reference to what struct that actually is in code! Note that the situation here is unchanged if other crates add custom features.

The Serde-recommended way to work around this is replacing the `Feature` trait with an enum:

```rust
enum Feature {
  AtomFeature(...),
  MirrorFeature(...),
  ...
}

impl Feature {
  fn apply(&self, ...) -> ... {
    match self {
      AtomFeature(coords, etc) => // code for applying an atom feature,
      // ... etc
    }
  }
```

This would make serialization and deserialization trivial, but is badly factored. If we expect to have ~100 features, like most major CAD programs seem to, then we need a 100-arm match statement! And each of the arms is likely to be ~80LOC + a bunch of helper function calls. This is a readability nightmare, but could be made stomachable using macros or something. Also, now we can't just add a method to the `Feature` trait - we have to introduce a match-monster like the above whenever we want to do something that varies feature to feature.

Another shortcoming of this approach is that features cannot be added by other crates. Right now, we have a PdbFeature inside the pdb crate. To use the enum approach, we'd have to put the PdbFeature `apply` code in `scene`, and introduce an uncomfortable dependency. I don't think the scene graph ought to worry about the protein data bank.

This leaves us with a few solutions that I can think of. First:
- We introduce an enum `FeatureValue` that can contain any of the common types a `Feature` uses (`AtomSpecifier`, int types, `String`, `FeatureId`, etc). `FeatureValue` implements `serde::Serialize` and `serde::de::DeserializeOwned`.
- `Feature` remains a trait.
  + It does not implement `serde::Serialize`, because that trait uses a generic method (and so if Feature was Serialize, dyn Feature would violate rust's object safety rules).
  + It requires `Feature::serialize(&self) -> HashMap<String, FeatureValue>`, which writes all the Feature's data out as a map. This map must also include a string that represents the feature's name under the key "name"
- The main crate builds a global feature registry at runtime. (`HashMap<String, Fn(HashMap<String, FeatureValue>) -> Result<dyn Feature, FeatureDeserializationError>>`).
- `FeatureList` implements custom serialization and deserialization.
  + To serialize, it maps the `Vec<dyn Feature>` into a `Vec<HashMap<String, FeatureValue>>` using `Feature::serialize`. It then stores the feature data in that trivially serializable form.
  + To deserialize, it does the opposite: deserializing first to `Vec<HashMap<String, FeatureValue>>`, then mapping the vec as such:
    * Get the `name` value from the hashmap form
    * Use the registry to get the function that is responsible for creating those features
    * pass the hashmap into that function and get the `dyn Feature` it returned to you.
This is a complicated system, but it would work. Additionally, it provides a runtime feature registry, which could useful for building the UI or implementing a feature search. It is about as extensible as the feature system could get, I think.

One downside of this implementation is that it is relatively slow (but probably still msec scale and happens infrequently). It is also somewhat memory intensive - running `Feature::serialize` is effectively a clone, but to be fair the resulting maps can be dropped as soon as they are written to disk. Finally, it.. *feels squishy?* My gut says it's bad to pass a struct around as a hashmap, because now you have to error handle things when the keys are wrong or missing, or if the value isn't the right type. That problem is inherent to serialization, though. Either Serde worries about those mistakes or we do, and in this case it seems like our code is better postured to handle this.

The second approach:
- `Feature` is an enum, not a trait. It enumerates the core features that you need to edit the scene, but also provides extensibility for storing an `ExternalFeature` (`Feature::ExternalFeature(ExternalFeature)`)
- `ExternalFeature` implements custom serialization and deserialization using a similar approach to the one just described (global feature registry + `HashMap<String, FeatureValue>`).
This initially had my favour - it lets us handle core features easily, because Serde loves Enums. It also lets us sweep this whole problem down the timeline, because we can just comment out `ExternalFeature` until we really need to address the issue.

However, this solution still demands all the complexity and tradeoffs as the previous approach. Additionally, it creates different codepaths for 'built-in' and 'external' features, whereas the previous system treated them totally homogeneously. As such, I now think this is a bad approach.

The final approach is kind of a bummer, but might be the most sane of the three:
- `Feature` is an enum.
- We use macros to improve code organization (to kill the match-monster)
- We simply do not allow other crates or modules to add new features.
This is very simple (and works well) but highly restrictive. To be fair, it is not as much of a wet blanket as it sounds like: we can still add custom features via the modelling language, because it's easy to just add one enum case for all of them (i.e. `Feature::CustomFeature(String)`, where the string is the modelling-lang script).

It also requires us to move primitive imports like `PdbFeature` into the scene crate and treat them as first class.

I think I'm leaning towards that final approach. @maaku, I'd appreciate your thoughts on this!

**Comment by shinzlet** [comment]
I'm in a bit of a time crunch before school begins, so I'm opting to take the third approach (slightly worse code organization in exchange for much stronger / faster serialization using a Feature enum). This can be changed in the future relatively easily so it is not worth being a blocker.
